larpandoracontent/LArHelpers/LArGeometryHelper.cc

```diff
@@ -422,6 +422,58 @@ bool LArGeometryHelper::IsInGap3D(const Pandora &pandora, const CartesianVector
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
+float LArGeometryHelper::ProjectAcrossGap3D(const Pandora &pandora, const CartesianVector &testPoint, const CartesianVector &testDirection,
+        const float gapTolerance, const int recurseLimit)
+{
+    float bestGapSize = std::numeric_limits<float>::max();
+
+    for (const DetectorGap *const detectorGap : pandora.GetGeometry()->GetDetectorGapList())
+    {
+        if (!detectorGap->IsInGap(testPoint, TPC_3D, gapTolerance))
+            continue;
+
+        const LineGap* lineGap = dynamic_cast<const LineGap *>(detectorGap);
+
+        if (lineGap == nullptr)
+            continue;
+
+        // ATTN: Right now, this only works for gaps with only X components.
+        //       If the Z is used for detector gaps, can cause infs.
+        const float gapWidth = std::fabs(lineGap->GetLineStartX() - lineGap->GetLineEndX());
+
+        const float distanceToStart = std::fabs(testPoint.GetX() - lineGap->GetLineStartX());
+        const float distanceToEnd = std::fabs(testPoint.GetX() - lineGap->GetLineEndX());
+        const float middleOfGap = std::fabs((lineGap->GetLineStartX() + lineGap->GetLineEndX()) / 2);
+        const CartesianVector gapPoint(middleOfGap, 0, testPoint.GetZ());
+
+        const float distanceToGap = std::min(distanceToStart, distanceToEnd);
+        const float totalGapSize = distanceToGap + gapWidth;
```

Comment:
It is not clear to me if this is the best X range over which to project, in case the point is inside the gap...

Reply:
I've updated this now, to calculate the distance from the current point to each gap edge, and then pick the most appropriate edge based on the given direction.

I did a quick local test for the 6 cases (left / inside / right of gap, going left/right) and got the expected answers out each time (1/2 gap width when in the middle of the gap, full gap plus threshold either side and pointing towards it, 0 when pointing away from the gap).

I'll do a full scale check of this on my sample when I've implemented the other comments (as well as inside the CI) to verify everything is working as expected though.